### ISSUE TITLE
Recheck periodically for better quality videos

### ISSUE THREAD

**Issue by ankenyr** [tubearchivist/tubearchivist#265]
This being an archiving project, downloading the best quality desired by the [Format Scheme](https://github.com/tubearchivist/tubearchivist/wiki/Settings#download-format) is presumably desired. If a user wants 4k but only 2k is available, 2k is what will be downloaded.

Sometimes however the 4k transcode can be delayed and not available for some time. [In certain cases it is possible for this to take even weeks](https://github.com/TheFrenchGhosty/TheFrenchGhostys-Ultimate-YouTube-DL-Scripts-Collection/blob/master/docs/Scripts-Type.md#archivist-scripts)

It would be good to emulate what TheFrenchGhosty is doing where you immediately download what ever the best quality is right as soon as it is available. Then periodically check for higher quality videos to be available. This periodic check only happens for videos that are younger than the cut-off period in order to not waste time checking videos that are too old. This period should be configurable by the user.

**Comment by bbilly1** [comment]
Thanks for reaching out on Discord and thanks for adding the links here. We already have the periodic rescan task, I think we can add this there. 

But first step would be to index the file metadata into Elasticsearch so that can be compared on the next scan.

**Comment by ankenyr** [comment]
Something that will be difficult is how to determine if something is more desired based on the format selection string. From my understanding the selection language can be quite complex but at least for FrenchGhost's there is a list of items separated by `/`

Will write up a few edge cases with my suggestion in a tl;dr on how I think the feature should work below

Lets say you have an video stored and its metadata. You have your format select string and a new video is downloaded. You can try and compare the videos but that isn't always straight forward. I may prefer H265 over VP9. So straight media to media comparisons isn't the best.

You could just take what ever the new thing is if it is different but what happens if the content creator removed a 4k video for some reason. You have 4k and now a 2k is downloaded. Or similarly with codecs.

You could try and parse the selection format string but that is fraught with issues itself. What happens if the format is changed? How complex is it really? I suspect you can do a lot of crazy stuff with it.

My suggestion would be to allow someone to make multiple selections. For me I would break up FrenchGhosts script into the 102 different individual selections such as 
`bestvideo[vcodec^=av01][height>=4320][fps>30`
`bestvideo[vcodec^=vp9.2][height>=4320][fps>30]`
`bestvideo[vcodec^=vp9][height>=4320][fps>30]`
being the first three selections. These would be in their priority. TA would perform a request for each one in order till one is successful and then store which index (SI) was successful. Subsequent runs would only attempt index 0 to SI-1. If a new video is found it updates SI and the media.

We should also reach out to the dev of yt-dlp, they are super responsive and may have an even better idea that I am not thinking of on this.

**Comment by bbilly1** [comment]
@ankenyr are you still interested in working on this? v0.3.6 now indexes all required metadata, first required step towards checking if there is a better resolution out there.

**Comment by CastleBravoMegatons** [comment]
I am going to tackle this. Just a heads up we are probably going to need a migration command as FPS was not being collected and put into the Database

**Comment by bbilly1** [comment]
We are about to start our refactor and frontend migration. You'll probably want to wait with that, at least until the backend refactor is done then fork off from that. Otherwise merging will be difficult.

Also see CONTRIBUTING.md [here](https://github.com/tubearchivist/tubearchivist/blob/master/CONTRIBUTING.md#next-steps).

List of issues [here](https://github.com/tubearchivist/tubearchivist/issues?q=is%3Aissue+is%3Aopen+label%3A%22react+migration%22).

**Comment by MerlinScheurer** [comment]
This needs to take into account that you might not want to re-download if the length of the video changed.
If an creator removed a chunk of the video, for example.

**Comment by bbilly1** [comment]
Yeah, there is always a risk of unintended consequences, for example copyright claims can result in missing audio or partially black screen during the disputed timestamps. 

Ultimately well need to see how automated we can do this.

Or somehow keep multiple versions of the same video...

**Comment by Handrail9** [comment]
> Yeah, there is always a risk of unintended consequences, for example copyright claims can result in missing audio or partially black screen during the disputed timestamps.
> 
> Ultimately well need to see how automated we can do this.
> 
> Or somehow keep multiple versions of the same video...

I am not a developer with enough experience to say if this would be viable, but if I was I would try to do something like

- Add a settings toggle for users to automatically check for higher quality (newer?) versions of a video
- Sub-option to ask the user if they want to keep both version of a video
- Sub-option to ask the user if they want to be notified of a new version of a video (Via UI popup on the home screen, web browser push, UnifiedPush (ntfy), or email)

Some other things may be able to be used like comparing the video length and notifying a user, or hashing the audio and seeing if the audio file is the same (if that is possible, once again, not a dev I could be blowing hot air)

If this is added I was hoping to use it with YT Premium to get the higher bitrate 4k files. I don't have Premium atm but I was going to see if I could grab it for like a month and then "upgrade" the videos I have.

**Comment by IAmABakuAMA** [comment]
I was actually just thinking on the viability of this last night, as I'd really like to see it implemented and was wondering how tricky it would be to implement.

Something else to consider is that it also needs to double-check the max quality set for that channel, or the program as a whole if no channel override is set. It would be incredibly annoying to be told that 4k upgrade is available when most channels in my collection have a configured max quality of 1080p.

I think fetching and comparing video length pre download is the best that could be done, in terms of not overwriting a changed video. That would prevent videos that have been trimmed down from replacing the full length video. But that misses the obvious edge cases of black bars being placed over parts of the video, or audio being bleeped or muted. I don't think hashing is possible though, as I'm pretty sure audio formats also change after upload, or at least with different video qualities.

Perhaps a double layered approach is the best that could be done:
- Compare video lengths, and if different, don't download
- Check the description to see if it's been copyright claimed, which could result in muted audio, and if so, either don't download, or add an option in settings to ignore and download anyway

But that still won't be perfect, and will still inevitably lead to lost data. I doubt the project can support having multiple copies of the same video without some extensive alterations.

That said, I feel it is worth pointing out that our current approach for comment handling is to fetch them all, and then delete all current comments, replacing them with whatever what redownloaded. There is an issue opened for this, so it may be fixed eventually, but I don't think the project aims for "100% archival", rather "good enough" archival. I think it would be worth doing anyway, even if all it does is compare the video lengths.
